### ISSUE TITLE
String/bytes issue in cython with Python3

### ISSUE THREAD

**Issue by bmoscon** [manahl/arctic#64]
The cython code has some issues with Python3. It assumes input is in bytes, not string format, and it returns things in bytes format (not string). Its easy to fix this, you can use this directive:

cython: c_string_type=str, c_string_encoding=ascii

to tell it to expect str format and you can cast the return from decompress to a c string from bytes. However, this breaks compatibility with the lz4 code, which continues to return things in bytes. 

So if you then decide to cast things to bytes, and return them in bytes you have a different issue in that the unit tests are all in strings, and so everything fails. 

There are several ways to fix/work around this, but I'm not sure what is best. Also, the 
tests in tests/unit/test_compress seem to call directly into the cython code. When does the wrapper in arctic/_compression.py ever get used? That could be a place to address this since in there you can do all the UTF-8 encoding and decoding without needing to modify any other code. 

Let me know what you think @TomTaylorLondon @jamesblackburn

**Comment by TomTaylorLondon** [comment]
arctic/_compression.py handles switching between our cython parallel lz4 and the single-threaded version. In the short term we could modify artic/_compression.py to only use the cython in 2.7.

Longer term does this help:
http://stackoverflow.com/questions/22487780/what-do-i-use-instead-of-pystring-asstring-when-loading-a-python-module-in-3-3

**Comment by TomTaylorLondon** [comment]
Oh and in regards to the tests - there are two sets:
test_compress.py - test the cython
test_compression.py - test the python interface

**Comment by bmoscon** [comment]
@TomTaylorLondon that stack overflow post is more about a specific API that was removed in the python C API in python3. I've already fixed that issue. This issue is more that we can easily control everything in the cython code, but the LZ4 code is straight C, so we have less control over that. So, whats the best way to fix it? I'll play with it some more tomorrow and see if I can find a solution I think is best

**Comment by bmoscon** [comment]
@TomTaylorLondon Per our conversation I did some more testing. In Python3, the LZ4 lib arctic uses ALWAYS returns the PYthon3 unicode byte array. Most the code in Arctic is set up to work on normal Python2 strings. I've changed the Cython code to return Python2 strings, but that doesn't solve the LZ4 library issue. Either arctic will need to drop support for strings and expect byte arrays, or we'll need to drop support for LZ4, or even worse, provide our own LZ4 library that does what we need.
